### PR TITLE
(HTCONDOR-2431)  $(epoch-type)_JOB_ATTRS

### PR DESCRIPTION
**Pull request by Todd-L-Miller**
Implemented in a generic way because it didn't take any longer than being specific.

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [x] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [x] Verify that the branch destination of the PR matches the target version of the ticket
- [x] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed  (documentation [build logs](https://readthedocs.org/projects/htcondor/builds/))
- [x] Check for version history, if needed
- [x] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: src/condor_utils/param_info.in [hunk 1]
```diff
@@ -1135,6 +1135,10 @@ default=
 type=string
 tags=had,AbstractReplicatorStateMachine
 
+[TRANSFER_JOB_ATTRS]
+default=ClusterID, ProcID, NumShadowStarts
+types=string
```
Comment (ColeBollig) [inline src/condor_utils/param_info.in:1140]:
```suggestion
type=string
```

File: src/condor_utils/job_ad_instance_recording.cpp [hunk 2]
```diff
@@ -107,6 +107,36 @@ initJobEpochHistoryFiles(){
...
+copyEpochJobAttrs( const classad::ClassAd * job_ad, const classad::ClassAd * other_ad, const char * banner_name ) {
+    std::string paramName;
+    formatstr( paramName, "%s_JOB_ATTRS", banner_name );
+
+    // For admins to explicitly specify no attributes, these three
+    // parameters must NOT be in the param table.
+    if(! param_defined_by_config(paramName.c_str())) {
+        if( (strcmp(banner_name, "INPUT" ) == 0) ||
+          (strcmp(banner_name, "OUTPUT" ) == 0) ||
+          (strcmp(banner_name, "CHECKPOINT" ) == 0) ) {
```
Comment (ColeBollig) [inline src/condor_utils/job_ad_instance_recording.cpp:121]: I'm cool if you keep this as is, but I personally have a preference for using the `MATCH` macro for `strcmp()`.

Comment (Todd-L-Miller) [inline src/condor_utils/job_ad_instance_recording.cpp:121]:
```suggestion
        if( (strcmp(banner_name, "INPUT" ) == MATCH) ||
          (strcmp(banner_name, "OUTPUT" ) == MATCH) ||
          (strcmp(banner_name, "CHECKPOINT" ) == MATCH) ) {
```

**Review by ColeBollig — APPROVED** [review]
LGTM

**Pull request merged**